onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+  // Iterate through all subscript labels in the output subscript
+  bool is_in_middle_of_ellipsis = false;
+  int64_t ellipsis_char_count = 0;
+
+  subscript_indices_to_output_indices_.resize(num_subscript_indices_, -1);
+
+  std::array<int64_t, EinsumOp::num_of_letters> output_letter_to_count;
+  output_letter_to_count.fill(0);
+
+  // Arbitrarily reserve some space as we don't expect rank of output to be > 10 (pay re-allocation cost if it is)
+  output_dims_.reserve(10);
+
+  int64_t output_dim_counter = 0;
+  for (auto subscript_label : einsum_equation_preprocessor_.right_equation_) {
+    if (subscript_label == '.') {
+      is_in_middle_of_ellipsis = true;
+      // Make sure there aren't more than 3 '.'s in the current subscript
+      ORT_ENFORCE(++ellipsis_char_count <= 3, "Found a '.' not part of an ellipsis in the output");
+      if (ellipsis_char_count == 3) {  // Ellipsis is complete. Process it.
+        is_in_middle_of_ellipsis = false;
+        for (size_t i = 0; i < num_of_ellipsis_dims_; ++i) {
+          output_dims_.push_back(subscript_indices_to_dim_value_[i]);
+          // The ellipsis is seen in the output and hence the corresponding dims are to not be reduced
+          subscript_indices_to_last_input_[i] = -1;
+          subscript_indices_to_output_indices_[i] = output_dim_counter++;
+        }
+      }
+    } else {
+      ORT_ENFORCE(!is_in_middle_of_ellipsis, "Found '.' not part of an ellipsis in the output");
+
+      ORT_ENFORCE(subscript_label >= 'a' && subscript_label <= 'z',
+                  "The only subscript labels allowed are lowercase letters (a-z)");
+
+      auto letter_index = static_cast<int64_t>(subscript_label - 'a');
+
+      ORT_ENFORCE(output_letter_to_count[letter_index] == 0,
+                  "Output subscript contains repeated letters");
+      ++output_letter_to_count[letter_index];
+
+      auto mapped_index = letter_to_index_[letter_index];
+      ORT_ENFORCE(mapped_index != -1,
+                  "Output subscript contains letters not seen in the inputs");
+
+      output_dims_.push_back(subscript_indices_to_dim_value_[mapped_index]);
+
+      // Reset the last input index for this subscript label
+      // given that it is seen in the output and hence can't be reduced
+      subscript_indices_to_last_input_[mapped_index] = -1;
+
+      subscript_indices_to_output_indices_[mapped_index] = output_dim_counter++;
+    }
+  }
+}
+
+void EinsumComputePreprocessor::PreprocessInputs() {
+  preprocessed_inputs_.reserve(inputs_.size());
+  // As part of input preprocessing we "homogenize" them by
+  // 1) Making them all of the same rank
+  // 2) The axes order in all the inputs are to be made the same
+  int64_t input_iter = 0;
+  for (const auto* input : inputs_) {
+    // We need to make a copy of the op's inputs because they will be mutated along the op's compute steps
+    Tensor preprocessed = input->Clone(allocator_);
+    const auto& input_dims = input->Shape().GetDims();
```

Comment:
Do we have to clone at this point? If we only need to change the shape and not do a transpose could we just construct a Tensor that points to the original data instead of copying the data?

Reply:
Fixed. Avoided cloning, code becomes a bit more esoteric but avoids the copy from being made up-front.